grpc-api/src/main/java/io/deephaven/grpc_api/browserstreaming/BrowserStream.java

```diff
@@ -0,0 +1,263 @@
+import io.deephaven.base.verify.Assert;
+import io.deephaven.grpc_api.session.SessionState;
+import io.deephaven.grpc_api.util.GrpcServiceOverrideBuilder;
+import io.deephaven.grpc_api.util.GrpcUtil;
+import io.deephaven.internal.log.LoggerFactory;
+import io.deephaven.io.logger.Logger;
+import io.grpc.StatusRuntimeException;
+import io.grpc.stub.StreamObserver;
+
+import java.io.Closeable;
+
+public class BrowserStream<T> implements Closeable {
+    public enum Mode {
+        /** Messages must be processed in order, if a gap is observed in sequences wait until the missing message arrives. */
+        IN_ORDER,
+        /** Always process the current message if it has the highest sequence, if an old message arrives, ignore it. */
+        MOST_RECENT
+    }
+
+    /**
+     * Creates a BrowserStream based on the current session and the observed passed in to the open stream call.
+     */
+    public interface Factory<ReqT, RespT> {
+        BrowserStream<ReqT> create(SessionState sessionState, StreamObserver<RespT> responseObserver);
+    }
+
+    private static class Message<T> {
+        private int pos;
+        private final T message;
+        private final StreamData streamData;
+
+        public Message(T message, StreamData streamData) {
+            this.message = message;
+            this.streamData = streamData;
+        }
+
+        public T getMessage() {
+            return message;
+        }
+
+        public StreamData getStreamData() {
+            return streamData;
+        }
+    }
+
+    private interface Marshaller<T> {
+        void onMessageReceived(T message);
+        void onCancel();
+        void onError(Throwable err);
+        void onCompleted();
+    }
+
+    private static final Logger log = LoggerFactory.getLogger(BrowserStream.class);
+
+    /**
+     * Builds a BrowserStream factory based on the given mode and bidirectional stream method.
+     */
+    public static <ReqT, RespT> Factory<ReqT, RespT> factory(Mode mode, GrpcServiceOverrideBuilder.BidiDelegate<ReqT, RespT> bidiDelegate) {
+        return (session, responseObserver) -> new BrowserStream<>(mode, session, new Marshaller<ReqT>() {
+            private final StreamObserver<ReqT> requestObserver = bidiDelegate.doInvoke(responseObserver);
+            @Override
+            public void onMessageReceived(ReqT message) {
+                requestObserver.onNext(message);
+            }
+
+            @Override
+            public void onCancel() {
+                StatusRuntimeException canceled = GrpcUtil.statusRuntimeException(Code.CANCELLED, "Stream canceled on the server");
+                GrpcUtil.safelyExecute(() -> {
+                    responseObserver.onError(canceled);
+                });
+
+                //TODO is this necessary or does the above result in this being called automatically?
+                GrpcUtil.safelyExecute(() -> {
+                    requestObserver.onError(canceled);
+                });
```

Comment:
If this is truly the onCancel from the ServerStream observer, then you will be unable to send a message to the observer. In either case this onError call is duplicated.

Reply:
onCancel() is only called by close(), which can only happen on session expiry (or possibly explicitly close of the rpc ticket? that would be weird imho...). Instead we should probably rename this method (and test as the TODO says)?